### PR TITLE
apply: fix witness raft log gc panic and refactor

### PR DESCRIPTION
**Pull request by ethercflow**
ref tikv/tikv#12876

Signed-off-by: Wenbo Zhang <ethercflow@gmail.com>

### What is changed and how it works?

Issue Number: ref #12876 

What's Changed:

- Fix panic caused by not initialize `apply_ctx.timer`, the `ApplyPoller` releases the ownership of `ApplyFsm` after skip flushing, and the other `ApplyPoller` gets this `ApplyFsm` then updates `applied_index` and `flush` first, causing the applied index to flip.
- Refactor the logic of notifying the raftstore thread. In the previous method, after calling `finish_for` (which will update the `apply_ctx.applied_res` array), explicitly call `notify_one` to notify `raftstore`, resulting in redundant notifications. Now the notification is done uniformly through `apply_ctx.applied_res`.
- Added a separate configuration to control the interval of request voter_replicated_index

```commit-message
fix witness raft log gc panic and refactor
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:

### Check List 

Tests 

- Integration test
- Manual test (add detailed scripts or steps below)
  - Setup cluster use sysbench prepare as workload,  turn down request voter_replicated_index interval, keep running for more than 1h. (When there was a problem before, it can be reproduced in a few minutes)

Side effects

- Performance regression
    - None

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/raftstore/src/store/fsm/apply.rs [hunk 1]
```diff
@@ -4109,18 +4104,29 @@ where
...
                     ctx.prepare_for(&mut self.delegate);
-                    self.delegate.write_apply_state(ctx.kv_wb_mut());
-                    ctx.commit_opt(&mut self.delegate, true);
-                    ctx.finish_for(&mut self.delegate, VecDeque::new());
-                    ctx.notifier
-                        .notify_one(self.delegate.region_id(), PeerMsg::ApplyRes { res });
+                    let mut result = VecDeque::new();
+                    // If modified `truncated_state` in `try_compact_log`, the apply state should be
+                    // persisted.
+                    if should_write {
```
Comment (Connor1996) [inline components/raftstore/src/store/fsm/apply.rs:4124]: Maybe use `ctx.kv_wb_mut().is_empty()`

Comment (ethercflow) [reply components/raftstore/src/store/fsm/apply.rs:4124]: We update `WB` after this check, and `WB` may be empty before that, so I think `should_write` is needed.

File: components/raftstore/src/store/fsm/apply.rs [hunk 2]
```diff
@@ -688,20 +689,21 @@ where
...
 
     /// Flush all pending writes to engines.
-    /// If it returns true, all pending writes are persisted in engines.
-    pub fn flush(&mut self) -> bool {
-        // TODO: this check is too hacky, need to be more verbose and less buggy.
+    pub fn flush(&mut self) {
         let t = match self.timer.take() {
             Some(t) => t,
-            None => return false,
+            // This will lead to inaccurate delay calculation of write requests with uninitialized
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/apply.rs:695]: This is a bad idea. It will cause confusion for coprocessors if there is no writes at all.

Comment (Connor1996) [reply components/raftstore/src/store/fsm/apply.rs:695]: but timer.is_none() standing for no write is confusing as well.

File: components/raftstore/src/store/fsm/apply.rs [hunk 3]
```diff
@@ -2970,7 +2971,7 @@ where
         &mut self,
         voter_replicated_index: u64,
         voter_replicated_term: u64,
-    ) -> Result<Option<TaskRes<EK::Snapshot>>> {
+    ) -> Result<(bool, Option<ExecResult<EK::Snapshot>>)> {
```
Comment (BusyJay) [inline components/raftstore/src/store/fsm/apply.rs:2974]: Should comment what's the meaning of returned bool.

**Review by BusyJay — APPROVED** [review]
Rest LGTM

File: components/raftstore/src/store/fsm/apply.rs [hunk 4]
```diff
@@ -300,7 +301,7 @@ pub enum ExecResult<S> {
     SetFlashbackState {
         region: Region,
     },
-    PendingCompactCmd,
+    HasPendingCompactCmd(bool),
```
Comment (tonyxuqqi) [inline components/raftstore/src/store/fsm/apply.rs:304]: Please add a comment for what's the bool's meaning.

**Review by tonyxuqqi — APPROVED** [review]

**Pull request merged**